PhysicsTools/PatAlgos/src/MuonMvaEstimator.cc

```diff
@@ -0,0 +1,139 @@
+#include "PhysicsTools/PatAlgos/interface/MuonMvaEstimator.h"
+#include "DataFormats/MuonReco/interface/MuonSelectors.h"
+#include "DataFormats/VertexReco/interface/Vertex.h"
+#include "DataFormats/JetReco/interface/PFJet.h"
+#include "DataFormats/JetReco/interface/PFJetCollection.h"
+#include "DataFormats/Candidate/interface/Candidate.h"
+#include "JetMETCorrections/JetCorrector/interface/JetCorrector.h"
+
+using namespace pat;
+
+MuonMvaEstimator::MuonMvaEstimator():
+  tmvaReader_("!Color:!Silent:Error"),
+  initialized_(false),
+  mva_(0),
+  dRmax_(0)
+{}
+
+void MuonMvaEstimator::initialize(std::string weightsfile,
+				  float dRmax)
+{ 
+  if (initialized_) return;
+  tmvaReader_.AddVariable("LepGood_pt",                    &pt_               );
+  tmvaReader_.AddVariable("LepGood_eta",                   &eta_              );
+  tmvaReader_.AddVariable("LepGood_jetNDauChargedMVASel",  &jetNDauCharged_   );
+  tmvaReader_.AddVariable("LepGood_miniRelIsoCharged",     &miniRelIsoCharged_);
+  tmvaReader_.AddVariable("LepGood_miniRelIsoNeutral",     &miniRelIsoNeutral_);
+  tmvaReader_.AddVariable("LepGood_jetPtRelv2",            &jetPtRel_         );
+  tmvaReader_.AddVariable("min(LepGood_jetPtRatiov2,1.5)", &jetPtRatio_       );
+  tmvaReader_.AddVariable("max(LepGood_jetBTagCSV,0)",     &jetBTagCSV_       );
+  tmvaReader_.AddVariable("LepGood_sip3d",                 &sip_              );
+  tmvaReader_.AddVariable("log(abs(LepGood_dxy))",         &log_abs_dxyBS_    ); 
+  tmvaReader_.AddVariable("log(abs(LepGood_dz))",          &log_abs_dzPV_     );
+  tmvaReader_.AddVariable("LepGood_segmentCompatibility",  &segmentCompatibility_);
+  tmvaReader_.BookMVA("BDTG",weightsfile);
+  dRmax_ = dRmax;
+  initialized_ = true;
```

Comment:
what is the memory footprint of this?

Looking at the training file, it is BDT::BDTG. If I'm not mistaken, the evaluation can go via GBRForest.
If so, please migrate
An example is available e.g. in https://github.com/cms-sw/cmssw/blob/master/RecoJets/JetProducers/src/PileupJetIdAlgo.cc

Reply:
No clue. How would I estimate it? I would prefer to avoid changes in the algorithm without a detailed test. This can be done later when we finally have all samples and a new training is ready.

Reply:
If there is a chance memory is 20-30 MB, this will can become a problem to run in production.
(T0 mem footprint will go up by 200MB).
Checks can be done with igprof
e.g.
https://twiki.cern.ch/twiki/bin/viewauth/CMS/RecoIntegration#Run_profiler_igprof
see " Checking memory use with IgProfService "

Reply:
https://slava77sk.web.cern.ch/slava77sk/reco/cgi-bin/igprof-navigator/CMSSW_9_4_X_2017-09-28-1100-sign969.pat-from-aod.10224.IgProf.1.MEM_LIVE/1420
this shows just under 3MB
This is perhaps OK for now.
Still, please please put it on the list to move to using GBRForest for evaluation: the pattern of changes is simple, e.g.   https://github.com/cms-sw/cmssw/blob/master/RecoJets/JetProducers/src/PileupJetIdAlgo.cc 
It just takes a factor of a few less memory.

Reply:
3MB per thread? (@slava77)